tests/ui/consts/const-eval/issue-91827-extern-types.rs

```diff
@@ -28,7 +28,10 @@ pub struct ListImpl<T, const N: usize> {
 
 impl<T> List<T> {
     const fn as_slice(&self) -> &[T] {
-        unsafe { std::slice::from_raw_parts(self.data.as_ptr(), self.len) }
+        unsafe {
+            let ptr = (self as *const List<T>).cast::<usize>().add(1).cast::<T>();
+            std::slice::from_raw_parts(ptr, self.len)
```

Comment:
What is this even doing? It looks like the original code would run into the `&Header` issue, but that issue doesn't have a work-around in Stacked Borrows...

Reply:
CI is failing because I of padding that I failed to account for in my poorly-written, manual, `offset_of!`.

I don't entirely follow your reasoning, but I think the code is UB before we even get to where you are seeing the `&Header` pattern. The slice can't be constructed because the array is length 0:
```
error: Undefined Behavior: trying to retag from <3086> for SharedReadOnly permission at alloc1[0x8], but that tag does not exist in the borrow stack for this location
   --> /home/ben/rust/library/core/src/slice/raw.rs:100:9
    |
100 |         &*ptr::slice_from_raw_parts(data, len)
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |         |
    |         trying to retag from <3086> for SharedReadOnly permission at alloc1[0x8], but that tag does not exist in the borrow stack for this location
    |         this error occurs as part of retag at alloc1[0x8..0x38]
    |
    = help: this indicates a potential bug in the program: it performed an invalid operation, but the Stacked Borrows rules it violated are still experimental
    = help: see https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md for further information
help: <3086> would have been created here, but this is a zero-size retag ([0x8..0x8]) so the tag in question does not exist anywhere
   --> tests/ui/consts/const-eval/issue-91827-extern-types.rs:31:45
    |
31  |         unsafe { std::slice::from_raw_parts(self.data.as_ptr(), self.len) }
    |                                             ^^^^^^^^^^^^^^^^^^
```

Reply:
I added an even more horrifying implementation that passes
```
x run miri --args tests/ui/consts/const-eval/issue-91827-extern-types.rs --target=wasm32-unknown-unknown
```

Reply:
The empty array is the header, isn't it?
```rust
pub struct List<T> {
    len: usize,
    data: [T; 0],
    tail: Opaque,
}
```
is basically a variable-sized array.

Also doesn't `tail_offset` further down already compute the offset you need?

Reply:
Ah I entirely forgot that I [made &Header work with extern types](https://github.com/rust-lang/miri/blob/67f6daff45c18cbd2c56f5684d6e49c0d73d83f5/src/borrow_tracker/stacked_borrows/mod.rs#L840-L846), so what is what is happening here.

Reply:
I don't quite understand why this is done in the way it is. Is there any value for the test to not just express this as `let ptr = (&self.data) as *const T;`?

Reply:
`&` restricts provenance to that field, which has size 0, so the result reference is only valid for 0 bytes, which is not very useful.

But `addr_of!(self.data)` should work.